vendor/riscv/riscv-isa-sim/riscv/Proc.cc

```diff
@@ -37,33 +59,136 @@ st_rvfi Processor::step(size_t n, st_rvfi reference) {
+    if(this->which_trap & ((reg_t)1 << (isa->get_max_xlen() - 1))) { 
+      //Since spike steps two times to take an interrupt, we store the intr value to the next step to return with rvfi
+      this->next_rvfi_intr |= 1 << 0; //intr [0]
+      this->next_rvfi_intr |= 1 << 2; //interrupt [2]
+      this->next_rvfi_intr |= 0x3FF8 & ((this->which_trap & 0xFF) << 3); //cause[13:3]
+    } else{
+      rvfi.trap |= 1 << 0; //trap [0]
+      rvfi.trap |= 1 << 1; //exception [1]
+      rvfi.trap |= 0x1F8 & ((this->which_trap) << 3); //exception_cause [8:3]
+      //TODO:
+      //debug_cause     [11:9] debug cause
+      //cause_type      [13:12]
+      //clicptr         [14]  CLIC interrupt pending
+      this->next_rvfi_intr = rvfi.trap; //store value to return with rvfi.intr on the next step
+    }
+  }
+
+  uint64_t last_rd_addr = 0;
+  uint64_t last_rd_wdata = 0;
   bool got_commit = false;
   for (auto &reg : reg_commits) {
-      if ((reg.first >> 4) > 32) {
-          if ((reg.first >> 4) < 0xFFF) {
-            for (size_t i = 0; i < CSR_SIZE; i++) {
-                if (!rvfi.csr_valid[i]) {
-                    rvfi.csr_valid[i] = 1;
-                    rvfi.csr_addr[i] = reg.first >> 4;
-                    rvfi.csr_wdata[i] = reg.second.v[0];
-                    rvfi.csr_wmask[i] = -1;
-                    break;
-                }
-            }
+    if((reg.first & 0xf) == 0x4) { //If CSR
+      for (size_t i = 0; i < CSR_SIZE; i++) {
+          if (!rvfi.csr_valid[i]) {
+              rvfi.csr_valid[i] = 1;
+              rvfi.csr_addr[i] = reg.first >> 4;
+              rvfi.csr_wdata[i] = reg.second.v[0];
+              rvfi.csr_wmask[i] = -1;
+              break;
           }
       }
-      else {
-        // TODO FIXME Take into account the XLEN/FLEN for int/FP values.
-        rvfi.rd1_addr = reg.first >> 4;
-        rvfi.rd1_wdata = reg.second.v[0];
-        // TODO FIXME Handle multiple register commits per cycle.
-        // TODO FIXME This must be handled on the RVFI side as well.
+    }
+    else {
+      if (got_commit) {
+        last_rd_addr = reg.first >> 4;
+        last_rd_wdata = reg.second.v[0];
+        continue;
+      }
+      // TODO FIXME Take into account the XLEN/FLEN for int/FP values.
+      rvfi.rd1_addr = reg.first >> 4;
+      rvfi.rd1_wdata = reg.second.v[0];
+      // TODO FIXME Handle multiple register commits per cycle.
+      // TODO FIXME This must be handled on the RVFI side as well.
+      got_commit = true; // Only latch first commit
+    }
+  }
+  // popret(z) should return rd1_addr = 0 instead of the SP to match with the cv32e40s core
+  if (((this->get_state()->last_inst_fetched.bits() & MASK_CM_POPRET) == MATCH_CM_POPRET) ||
+      ((this->get_state()->last_inst_fetched.bits() & MASK_CM_POPRETZ) == MATCH_CM_POPRETZ)) {
+    rvfi.rd1_addr = 0;
+    rvfi.rd1_wdata = 0;
+  }
+
+  bool mem_access = false; // TODO: support multiple memory writes/reads
```

Comment:
Just flagging all new TODOs in review comments, for visibility.

Reply:
At the moment, I only support reporting one memory read or write over RVFI per instruction. To account for this, I only compare the "lowest" bits in the comparison module. 40S can report multiple memory operations per instruction, so this should be added to fully match the core, but I did not prioritize implementing this since it is not essential.